File: implementations/cpp/recurrence_relations.cpp
/**
 * @file recurrence_relations.cpp
 * @brief Reference implementations for solving linear recurrence relations.
 *
 * Provides a modular companion matrix exponentiation engine capable of solving
 * any order-k linear recurrence in O(k^3 log n) time, along with dynamic programming
 * baseline verifiers.
 * Verified with C++17 (-std=c++17 -O3 -Wall -Wextra -Werror).
 */

#include <iostream>
#include <vector>
#include <cstdint>
#include <cassert>
#include <stdexcept>

namespace dsa {

/**
 * @brief Square matrix of dimension K x K with modular arithmetic operations.
 */
class SquareMatrix {
public:
    size_t k;
    uint64_t mod;
    std::vector<std::vector<uint64_t>> mat;

    SquareMatrix(size_t dim, uint64_t m)
        : k(dim), mod(m), mat(dim, std::vector<uint64_t>(dim, 0)) {}

    static SquareMatrix identity(size_t dim, uint64_t m) {
        SquareMatrix res(dim, m);
        for (size_t i = 0; i < dim; ++i) {
            res.mat[i][i] = 1 % m;
        }
        return res;
    }

    SquareMatrix operator*(const SquareMatrix& other) const {
        assert(k == other.k && mod == other.mod);
        SquareMatrix res(k, mod);
        for (size_t i = 0; i < k; ++i) {
            for (size_t p = 0; p < k; ++p) {
                if (mat[i][p] == 0) continue;
                for (size_t j = 0; j < k; ++j) {
                    __uint128_t term = static_cast<__uint128_t>(mat[i][p]) * other.mat[p][j];
                    res.mat[i][j] = static_cast<uint64_t>((res.mat[i][j] + term) % mod);
                }
            }
        }
        return res;
    }

    SquareMatrix power(uint64_t exp) const {
        SquareMatrix result = identity(k, mod);
        SquareMatrix base = *this;
        while (exp > 0) {
            if (exp & 1) result = result * base;
            base = base * base;
            exp >>= 1;
        }
        return result;
    }
};

/**
 * @brief Solves an order-k linear recurrence in O(k^3 log n) time.
 * Recurrence: a_n = c[0]*a_{n-1} + c[1]*a_{n-2} + ... + c[k-1]*a_{n-k}
 *
 * @param c Coefficients of length k.
 * @param init Base cases: [a_{k-1}, a_{k-2}, ..., a_0] in descending order of index.
 * @param n Target index to compute.
 * @param mod Modulo for all calculations.
 * @return a_n mod modulo.
 */
inline uint64_t solve_linear_recurrence(
    const std::vector<uint64_t>& c,
    const std::vector<uint64_t>& init,
    uint64_t n,
    uint64_t mod
) {
    size_t k = c.size();
    if (k == 0 || init.size() != k) {
        throw std::invalid_argument("Coefficient and initial condition dimensions must match and be > 0");
    }

    // Base cases check
    if (n < k) {
        // init is arranged [a_{k-1}, a_{k-2}, ..., a_0]
        return init[k - 1 - n] % mod;
    }

    // Construct companion matrix M:
    // Row 0: c[0], c[1], ..., c[k-1]
    // Row 1 to k-1: Subdiagonal 1s
    SquareMatrix M(k, mod);
    for (size_t j = 0; j < k; ++j) {
        M.mat[0][j] = c[j] % mod;
    }
    for (size_t i = 1; i < k; ++i) {
        M.mat[i][i - 1] = 1 % mod;
    }

    // Power M^(n - k + 1)
    SquareMatrix M_pow = M.power(n - k + 1);

    // Multiply M_pow by init vector [a_{k-1}, a_{k-2}, ..., a_0]^T
    uint64_t ans = 0;
    for (size_t j = 0; j < k; ++j) {
        __uint128_t term = static_cast<__uint128_t>(M_pow.mat[0][j]) * (init[j] % mod);
        ans = static_cast<uint64_t>((ans + term) % mod);
    }
    return ans;
}

/**
 * @brief Dynamic programming baseline for linear recurrences: O(n * k).
 */
inline uint64_t solve_linear_recurrence_dp(
    const std::vector<uint64_t>& c,
    const std::vector<uint64_t>& init,
    uint64_t n,
    uint64_t mod
) {
    size_t k = c.size();
    if (n < k) return init[k - 1 - n] % mod;

    std::vector<uint64_t> dp(n + 1, 0);
    for (size_t i = 0; i < k; ++i) {
        dp[i] = init[k - 1 - i] % mod;
    }

    for (size_t i = k; i <= n; ++i) {
        uint64_t val = 0;
        for (size_t j = 0; j < k; ++j) {
            __uint128_t term = static_cast<__uint128_t>(c[j] % mod) * dp[i - 1 - j];
            val = static_cast<uint64_t>((val + term) % mod);
        }
        dp[i] = val;
    }
    return dp[n];
}

/**
 * @brief Fast Fibonacci computation modulo MOD in O(log n).
 * F_0 = 0, F_1 = 1, F_n = F_{n-1} + F_{n-2}.
 */
inline uint64_t fibonacci(uint64_t n, uint64_t mod) {
    if (n == 0) return 0;
    if (n == 1) return 1 % mod;
    std::vector<uint64_t> c = {1, 1};
    std::vector<uint64_t> init = {1, 0}; // [F_1, F_0]
    return solve_linear_recurrence(c, init, n, mod);
}

/**
 * @brief Fast Tribonacci computation modulo MOD in O(log n).
 * T_0 = 0, T_1 = 1, T_2 = 1, T_n = T_{n-1} + T_{n-2} + T_{n-3}.
 */
inline uint64_t tribonacci(uint64_t n, uint64_t mod) {
    if (n == 0) return 0;
    if (n == 1 || n == 2) return 1 % mod;
    std::vector<uint64_t> c = {1, 1, 1};
    std::vector<uint64_t> init = {1, 1, 0}; // [T_2, T_1, T_0]
    return solve_linear_recurrence(c, init, n, mod);
}

} // namespace dsa

int main() {
    std::cout << "Running Recurrence Relations C++17 unit tests..." << std::endl;
    constexpr uint64_t MOD = 1'000'000'007ULL;

    // Test 1: Fibonacci verification against DP baseline for n = 0 to 100
    for (uint64_t n = 0; n <= 100; ++n) {
        uint64_t fib_mat = dsa::fibonacci(n, MOD);
        std::vector<uint64_t> c = {1, 1};
        std::vector<uint64_t> init = {1, 0};
        uint64_t fib_dp = dsa::solve_linear_recurrence_dp(c, init, n, MOD);
        assert(fib_mat == fib_dp);
    }
    assert(dsa::fibonacci(0, MOD) == 0);
    assert(dsa::fibonacci(1, MOD) == 1);
    assert(dsa::fibonacci(10, MOD) == 55);

    // Test 2: Tribonacci verification
    for (uint64_t n = 0; n <= 80; ++n) {
        uint64_t trib_mat = dsa::tribonacci(n, MOD);
        std::vector<uint64_t> c = {1, 1, 1};
        std::vector<uint64_t> init = {1, 1, 0};
        uint64_t trib_dp = dsa::solve_linear_recurrence_dp(c, init, n, MOD);
        assert(trib_mat == trib_dp);
    }
    assert(dsa::tribonacci(3, MOD) == 2);
    assert(dsa::tribonacci(4, MOD) == 4);
    assert(dsa::tribonacci(5, MOD) == 7);

    // Test 3: Arbitrary order-4 recurrence test: a_n = 2*a_{n-1} + 3*a_{n-2} + a_{n-3} + 4*a_{n-4}
    {
        std::vector<uint64_t> c = {2, 3, 1, 4};
        std::vector<uint64_t> init = {5, 3, 2, 1}; // [a_3, a_2, a_1, a_0]
        for (uint64_t n = 0; n <= 50; ++n) {
            uint64_t val_mat = dsa::solve_linear_recurrence(c, init, n, MOD);
            uint64_t val_dp = dsa::solve_linear_recurrence_dp(c, init, n, MOD);
            assert(val_mat == val_dp);
        }
    }

    // Test 4: Ultra-large n execution (n = 10^18)
    {
        uint64_t big_n = 1'000'000'000'000'000'000ULL;
        uint64_t big_fib = dsa::fibonacci(big_n, MOD);
        // Ensure it executes instantaneously and yields a valid modular remainder
        assert(big_fib < MOD);
        std::cout << "  [Matrix Exponentiation Verification] F(10^18) mod (10^9+7) = " << big_fib << std::endl;
    }

    std::cout << "[PASS] All Recurrence Relations C++ unit tests passed." << std::endl;
    return 0;
}

File: implementations/python/recurrence_relations.py
"""
Linear Recurrence Relations and Matrix Exponentiation Engine.

Provides an exact companion matrix solver for order-k linear recurrences
operating in O(k^3 log n) time, verified against dynamic programming baselines.
"""

import unittest
from typing import List


def matrix_mult(
    A: List[List[int]], B: List[List[int]], mod: int
) -> List[List[int]]:
    """Multiplies two k x k matrices under modulo arithmetic."""
    k = len(A)
    C = [[0] * k for _ in range(k)]
    for i in range(k):
        for p in range(k):
            if A[i][p] == 0:
                continue
            for j in range(k):
                C[i][j] = (C[i][j] + A[i][p] * B[p][j]) % mod
    return C


def matrix_pow(A: List[List[int]], exp: int, mod: int) -> List[List[int]]:
    """Computes A^exp under modulo arithmetic via binary exponentiation."""
    k = len(A)
    result = [[1 if i == j else 0 for j in range(k)] for i in range(k)]
    base = [row[:] for row in A]
    while exp > 0:
        if exp & 1:
            result = matrix_mult(result, base, mod)
        base = matrix_mult(base, base, mod)
        exp >>= 1
    return result


def solve_linear_recurrence(
    coefficients: List[int], initial_terms: List[int], n: int, mod: int
) -> int:
    """
    Solves order-k recurrence: a_n = sum_{j=0}^{k-1} c[j] * a_{n - 1 - j}
    in O(k^3 log n) time.

    initial_terms are given in descending index order: [a_{k-1}, a_{k-2}, ..., a_0].
    """
    k = len(coefficients)
    if k == 0 or len(initial_terms) != k:
        raise ValueError("Coefficients and initial terms must have equal length > 0")

    if n < k:
        return initial_terms[k - 1 - n] % mod

    # Construct companion matrix M
    M = [[0] * k for _ in range(k)]
    for j in range(k):
        M[0][j] = coefficients[j] % mod
    for i in range(1, k):
        M[i][i - 1] = 1 % mod

    # M^(n - k + 1)
    M_pow = matrix_pow(M, n - k + 1, mod)

    ans = 0
    for j in range(k):
        ans = (ans + M_pow[0][j] * (initial_terms[j] % mod)) % mod
    return ans


def solve_linear_recurrence_dp(
    coefficients: List[int], initial_terms: List[int], n: int, mod: int
) -> int:
    """Dynamic programming verification baseline in O(n * k) time."""
    k = len(coefficients)
    if n < k:
        return initial_terms[k - 1 - n] % mod

    dp = [0] * (n + 1)
    for i in range(k):
        dp[i] = initial_terms[k - 1 - i] % mod

    for i in range(k, n + 1):
        total = 0
        for j in range(k):
            total = (total + coefficients[j] * dp[i - 1 - j]) % mod
        dp[i] = total
    return dp[n]


def fibonacci(n: int, mod: int = 1_000_000_007) -> int:
    """Computes n-th Fibonacci number in O(log n) time."""
    if n == 0:
        return 0
    if n == 1:
        return 1 % mod
    return solve_linear_recurrence([1, 1], [1, 0], n, mod)


def tribonacci(n: int, mod: int = 1_000_000_007) -> int:
    """Computes n-th Tribonacci number in O(log n) time."""
    if n == 0:
        return 0
    if n in (1, 2):
        return 1 % mod
    return solve_linear_recurrence([1, 1, 1], [1, 1, 0], n, mod)


class TestRecurrenceRelations(unittest.TestCase):
    def test_fibonacci_base_cases(self):
        self.assertEqual(fibonacci(0), 0)
        self.assertEqual(fibonacci(1), 1)
        self.assertEqual(fibonacci(2), 1)
        self.assertEqual(fibonacci(3), 2)
        self.assertEqual(fibonacci(10), 55)

    def test_tribonacci_base_cases(self):
        self.assertEqual(tribonacci(0), 0)
        self.assertEqual(tribonacci(1), 1)
        self.assertEqual(tribonacci(2), 1)
        self.assertEqual(tribonacci(3), 2)
        self.assertEqual(tribonacci(4), 4)
        self.assertEqual(tribonacci(5), 7)

    def test_cross_validation_dp(self):
        mod = 1_000_000_007
        c = [2, 3, 1]
        init = [4, 1, 1]
        for n in range(50):
            mat_res = solve_linear_recurrence(c, init, n, mod)
            dp_res = solve_linear_recurrence_dp(c, init, n, mod)
            self.assertEqual(mat_res, dp_res)

    def test_large_n(self):
        mod = 1_000_000_007
        # F(10^18) should compute in a fraction of a second
        res = fibonacci(10**18, mod)
        self.assertEqual(res, 209783453)


if __name__ == "__main__":
    unittest.main()
